### PR TITLE
Do not unregister requests when waitRequest times out.

### PR DESCRIPTION
**Pull request by hernando**
I have a loop that does something like this:

    lunchbox::Request<bool> request = ... // some call that returns a Request<bool>
    bool success;
    while (true)
    {
        try
        {
            success = request.wait(500);
            break;
        }
        catch (const lunchbox::FutureTimeout&)
        {
            ... // Do something that will allow the request to be granted.
        }
    }

It's been a surprise to find out that after the first timeout, the request is unregistered and any subsequent call to request.wait fails immediately.

Reading at requestHandler.cpp:100 I've found this:

        const bool requestServed = request->lock.set( timeout );
        if( requestServed )
            result = request->result;

        unregisterRequest( requestID_ );

This basically means that a timeout can only occur once and there's no way to retry. Changing the code to this:

       const bool requestServed = request->lock.set( timeout );
        if( requestServed )
        {
            result = request->result;
            unregisterRequest( requestID_ );
        }

works for me, but I wonder whether this can have an undesired side-effect on Equalizer.

### CODE REVIEW

**Comment by eile** [comment]
I'll have to check Collage/Equalizer before merging this.

**Comment by eile** [comment]
Note: The API doc agrees with your change. ;)

**Comment by hernando** [comment]
Can it be merged then?

**Comment by eile** [comment]
Due to the BGAS paper and sw catalog having a higher prio, I did not get to this yet.

**Comment by eile** [comment]
I need to check in Co/Eq for timeout waits.

**Comment by eile** [comment]
This will need 274c244 plus Co/Eq changes to be done correctly. Changes unfortunately the API :(

**Comment by eile** [comment]
Doh, 274c244 is also not correct.

**Comment by hernando** [comment]
I think this PR and Eyescale/Collage/269 are enough. I haven't found any usage of Request::wait or RequestHandler::waitRequest with a timeout in Equalizer.

**Comment by tribal-tec** [comment]
I have not followed everything regarding this change, but the API breakage seems not necessary. Adding the new function should be enough, and documenting the existing one that it has shortcomings/different behavior/bugs.

**Comment by hernando** [comment]
You're right, the API breakage is not really necessary but Stefan insisted on removing relinquish. I also think that providing an unregistration method breaks symmetry, because registration of the request is done by the client using the RequestHandler directly, but unregistration is done with the Request. But ffter all he's the project owner, and I don't have an strong opinion about it.

The main motivation for relinquish is that some functions in Collage create requests that need to be yield to other thread and they are passed as request IDs using a stream, these requests mustn't be waited at destruction.
Then there are other functions that simply want to abandon the request after a timeout and there having a method than does relinquish and unregister in a single call may be less error prone.

For the first case move semantics can be the alternative so you can return rvalue requests and operator<< can take a Request&&, however the use of request IDs is pervasive. In the end we decided to simply use the old API and undeprecate the registration method that returns a request by ID. For the second case there aren't many alternatives you either have relinquish, followed by unregister, or a single method that does both from the request.
It's not nice breaking the API, but I's probably safe to assume that this function was only used by Equalizer and Collage in these two particular cases, and keeping the relinquish method would imply complicating the implementation of Request for ghost legacy code.

File: doc/Changelog.md [hunk 1]
```diff
@@ -12,6 +12,10 @@
   getLibraryPaths: Fixes library path results for OSX and linux
 * [213](https://github.com/Eyescale/Lunchbox/pull/213):
   getExecutablePath: More sensible behaviour for OS X app bundles
+* [233]((https://github.com/Eyescale/Lunchbox/pull/233):
+  Fix RequestHandler::waitRequest to not unregister the request if it
+  times out.
+  Request::relinquished replaced by Request::unregister.
```
Comment (eile) [inline doc/Changelog.md]: Release 1.12 section?

Comment (eile) [inline doc/Changelog.md]: In fact bump in CMakeLists is not needed?

Comment (hernando) [inline doc/Changelog.md]: You're right. It was already done.

**Comment by eile** [comment]
Merged through f78040e

**Pull request closed without merging**